DataFormats/EgammaReco/interface/EgTrigSumObj.h

```diff
@@ -0,0 +1,68 @@
+#ifndef DataFormats_EgammaReco_EgTrigSumObj_h
+#define DataFormats_EgammaReco_EgTrigSumObj_h
+
+#include "DataFormats/EgammaReco/interface/SuperCluster.h"
+#include "DataFormats/EgammaReco/interface/ElectronSeed.h"
+#include "DataFormats/GsfTrackReco/interface/GsfTrack.h"
+
+#include "DataFormats/EgammaReco/interface/SuperClusterFwd.h"
+#include "DataFormats/EgammaReco/interface/ElectronSeedFwd.h"
+#include "DataFormats/GsfTrackReco/interface/GsfTrackFwd.h"
+
+#include <unordered_map>
+#include <string>
+
+namespace reco {
+  class RecoEcalCandidate;
+}
+
+namespace reco {
+  class EgTrigSumObj {
+  public:
+    EgTrigSumObj() : energy_(0.), pt_(0.), eta_(0.), phi_(0.), hasPixelMatch_(false) {}
+
+    EgTrigSumObj(float energy, float pt, float eta, float phi);
+    EgTrigSumObj(const reco::RecoEcalCandidate& ecalCand);
+
+    float energy() const { return energy_; }
+    float et() const { return pt(); }
+    float pt() const { return pt_; }
+    float mass() const { return 0.; }
+    float eta() const { return eta_; }
+    float phi() const { return phi_; }
+
+    const reco::SuperClusterRef& superCluster() const { return superCluster_; }
+    const reco::GsfTrackRefVector& gsfTracks() const { return gsfTracks_; }
+    const reco::ElectronSeedRefVector& seeds() const { return seeds_; }
+
+    void setSuperCluster(const reco::SuperClusterRef& sc) { superCluster_ = sc; }
+    void setGsfTracks(reco::GsfTrackRefVector&& trks) { gsfTracks_ = trks; }
+    void setSeeds(reco::ElectronSeedRefVector&& seeds);
+
+    bool hasVar(const std::string& varName) const;
+    float var(const std::string& varName, bool raiseExcept = true) const;
+    const std::unordered_map<std::string, float>& vars() const { return vars_; }
+    //varNames and varNamesStr are reasonably expensive functions and are more
+    //intended for debugging than normal use
+    std::vector<std::string> varNames() const;
+    std::string varNamesStr() const;
+    void setVar(std::string name, float value, bool overwrite = true);
+    void clearVars() { vars_.clear(); }
+
+  private:
+    float energy_;
+    float pt_;
+    float eta_;
+    float phi_;
+    bool hasPixelMatch_;
+    std::unordered_map<std::string, float> vars_;
```

Comment:
do we have any other examples of `unordered_map` streamed in CMSSW?
@cms-sw/core-l2 do we have good support for this in root?

Reply:
> do we have any other examples of `unordered_map` streamed in CMSSW?

It seems we do have a few
https://github.com/cms-sw/cmssw/blob/00360a80172c673ea88608a98f6c106b1fa5de12/DataFormats/BTauReco/interface/DeepBoostedJetFeatures.h#L57
https://github.com/cms-sw/cmssw/blob/00360a80172c673ea88608a98f6c106b1fa5de12/DataFormats/L1THGCal/interface/HGCalClusterT.h#L195-L196
https://github.com/cms-sw/cmssw/blob/00360a80172c673ea88608a98f6c106b1fa5de12/DataFormats/L1THGCal/interface/HGCalTowerMap.h#L31

> @cms-sw/core-l2 do we have good support for this in root?

Good question, @pcanal (@Dr15Jones) could you comment?

How many elements is the map expected to have?

Traditional alternative would be a sorted `std::vector<std::pair<std::string, float>>` (or separate `std::vector<std::string>` and `std::vector<float>`, those would allow to make `varNames()` trivial).

Reply:
> > do we have any other examples of `unordered_map` streamed in CMSSW?
> 
> It seems we do have a few

IIRC, DeepBoostedJetFeatures is effectively transient (although it is not declared as such): it is not saved.

I'm not familiar with the L1THGCal cases.

We've recently gone through a case with problems of streaming a usual `std::map` in https://github.com/cms-sw/cmsdist/pull/6388#issuecomment-725620932
If this data is important it may be worth to pick a different type, like a suggested `vector<pair`

Reply:
thanks for the spot, I had not considered this. I used std::unorderedmap simply for convenience but we get better data stability with a sorted std::vector then thats not a problem.

It will contain on average a maximum of ~about 20, possibly 30 entries I believe but I  can check.  

On the splitting it to std::string and std::vector, so I personally dislike having two vectors coupled to each other as in we need to ensure they are the same length, kept up to date etc. This is not overly difficult and I'm happy to go with yours and others suggestions on what would be the best practice to create a set of read only key,value pairs in a class.